File: Go/345. Reverse Vowels of a String.go
package main

import "unicode"

//goland:noinspection GoUnusedFunction
func reverseVowels(s string) string {
	length := len(s)

	if length <= 1 {
		return s
	}

	vowels := map[rune]struct{}{
		'a': {}, 'e': {}, 'i': {}, 'o': {}, 'u': {},
	}

	slice := []rune(s)
	left := 0
	right := length - 1

	for left < length {
		if left >= right {
			break
		}

		if _, exists := vowels[unicode.ToLower(slice[left])]; exists {
			for right > left {
				if _, exists := vowels[unicode.ToLower(slice[right])]; exists {
					slice[left], slice[right] = slice[right], slice[left]
					right--
					break
				}

				right--
			}
		}

		left++
	}

	return string(slice)
}

File: TypeScript/345. Reverse Vowels of a String.ts
// noinspection JSUnusedGlobalSymbols
const reverseVowels = (s: string): string => {
  const length = s.length

  if (length <= 1) {
    return s
  }

  const vowels = new Set(['a', 'e', 'i', 'o', 'u'])
  let left = 0
  let right = length - 1
  const arr = s.split('')

  while (left < length) {
    if (left >= right) {
      break
    }

    if (vowels.has(s[left].toLowerCase())) {
      while (right > left) {
        if (vowels.has(s[right].toLowerCase())) {
          arr[left] = s[right]
          arr[right] = s[left]
          right--
          break
        }

        right--
      }
    }

    left++
  }

  return arr.join('')
}
